sort_images.sh

```diff
@@ -0,0 +1,159 @@
+    echo "Error: The file, " "${a_file_name}"
+    echo "- is missing exif date or modify date metadata - so skipping this file"
+    echo "Continuing with next file ..."
+    continue # To stop with current iteration & move on to next item.
+  elif [ "${exif_date}" ]; then
+    # Change filesystem date to EXIF photo-taken date, EXIF:DateTimeOriginal.
+    # Given Format:  2018:04:03 21:31:41
+    # Wanted Syntax: [[CC]YY]MMDDhhmm[.SS] Wanted Format: 2015-09-02_07-09_0060.jpg
+    # 1. Replace ALL occurrences of : 	With: nothing
+    # 2. Replace 1st occurrence of space 	With: nothing
+    # 3. Build string by deleting last 2 char at end, concatting . & then
+    # concatting those last 2 char, e.g., abc.12 = abc + . + 12
+
+    date_for_date_change="${exif_date//:/}"
+    date_for_date_change="${date_for_date_change// /}"
+    # Trim last 2 chars to remove SS so in format: [[CC]YY]MMDDhhmm[.SS]
+    date_for_date_change="${date_for_date_change%??}.${date_for_date_change: -2}"
+  else # if [ "${modify_date}" ]; then
+    # Change filesystem date to modify_date, date:modify for date change.
+    # Given Format:  2018-10-09T18:42:41+00:00
+    # Wanted Syntax: [[CC]YY]MMDDhhmm[.SS] Wanted Format: 202002031806
+    # 1. Trim last 9 chars.
+    # 2. Remove all -.
+    # 3. Remove all T.
+    # 4. Remove all :.
+    date_for_date_change="${modify_date::-9}"
+    date_for_date_change="${date_for_date_change//-/}"
+    date_for_date_change="${date_for_date_change//T/}"
+    date_for_date_change="${date_for_date_change//:/}"
+  fi
+  touch -t "${date_for_date_change}" "${a_file_name}"
+
+  #----------------------------------------------------------------------
+  # Make Subdirectories
+  #----------------------------------------------------------------------
+  # $date_for_date_change is: 202002031806
+  # ${string:position:length}
+  year="${date_for_date_change:0:4}"
+  month="${date_for_date_change:4:2}"
+  
+  just_path=$(dirname "${a_file_name}")
+  just_filename=$(basename "${a_file_name}") # For path to move files into subdirectories
+
+  if [ "${safe_day_subdir_also}" == 'y' ]; then # Make year-month-day subdirectories.
+    day="${date_for_date_change:6:2}"
+    path_with_subdir_year_month_day="${just_path}/${year}/${month}/${day}"
+    mkdir -p "${path_with_subdir_year_month_day}"
+    new_dir_and_filename="${just_path}/${year}/${month}/${day}/${just_filename}"
+  else # [ "${safe_day_subdir_also}" == 'n' ]; then # Make year-month subdirectories.
+    path_with_subdir_year_month="${just_path}/${year}/${month}"
+    mkdir -p "${path_with_subdir_year_month}"
+    new_dir_and_filename="${just_path}/${year}/${month}/${just_filename}"
+  fi
+
+  mv "${a_file_name}" "${new_dir_and_filename}"
+  file_sort_counter="$((file_sort_counter+1))"
+done < <(find "${directory_path}" -maxdepth 1 -type f -name '*.jpg' -o -name '*.JPG' \
+    -o -name '*.gif' -o -name '*.GIF' -o -name '*.tif' -o -name '*.TIF' \
+    -o -name '*.png' -o -name '*.PNG)
```

Comment:
You are missing a closing single quote (`'`) after `*.PNG`.

Reply:
@kcharter Argggggghh. Thanks for spotting that.

Reply:
Looks good, and I tested the change and the script no long crashes with a syntax error.